Implement C++ ModuleDict

ejguan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#47707 Implement C++ ModuleDict**

Fixes #45896
- [x] API
* constructor(OrderedDict/vector of pair)
* clear()
* items()
* keys()
* pop(string key)
* update(OrderedDict/vector of pair/ModuleDict)
* values()
* Iterator: begin/end
* size()
* empty()
* contains(string key)
* clone()
* pretty_print(stream)
* operator [string key]
* at(string key)
- [x] Add tests

Differential Revision: [D24872641](https://our.internmc.facebook.com/intern/diff/D24872641)

ejguan committed:
Implement C++ ModuleDict

[ghstack-poisoned]

ejguan committed:
Update on "Implement C++ ModuleDict"


Fixes #45896

- [ ] Add tests

Differential Revision: [D24872641](https://our.internmc.facebook.com/intern/diff/D24872641)

[ghstack-poisoned]

ejguan committed:
Update on "Implement C++ ModuleDict"


Fixes #45896

- [ ] Add tests

Differential Revision: [D24872641](https://our.internmc.facebook.com/intern/diff/D24872641)

[ghstack-poisoned]

zou3519 commented on torch/csrc/api/include/torch/nn/modules/container/moduledict.h:
```diff
@@ -0,0 +1,185 @@
+
+  ModuleDictImpl() = default;
+
+  /// Constructs the `ModuleDict` from a list of string-Module paris.
```
nit: pairs

zou3519 commented on torch/csrc/api/include/torch/nn/modules/container/moduledict.h:
```diff
@@ -0,0 +1,185 @@
+    return clone;
+  }
+
+  /// `reset()` is empty for `ModuleDict`, since it does not have parameters of
+  /// its own.
+  void reset() override {}
```
@glaringlee how is reset() used in the C++ API? If it doesn't make sense to call reset() for ModuleDict, would it be better to raise an error?

glaringlee replied:
@zou3519 
Can not raise an error, since this is called in cloneable.h when cloning, even it is empty, it will be called. Leave it empty is fine.

glaringlee replied:
@zou3519 I reviewed a bit, clone() is also overridden in this impl, so reset() is actually useless. I think it is fine to just leave it as blank, this is kinda convention in c++ API impls.

zou3519 commented on torch/csrc/api/include/torch/nn/modules/container/moduledict.h:
```diff
@@ -0,0 +1,185 @@
+    return module;
+  }
+
+  /// Updated the `ModuleDict` with key-value paris from vector of key-module pairs
```
nit: paris -> pairs

zou3519 commented on test/cpp/api/moduledict.cpp:
```diff
@@ -0,0 +1,304 @@
+  }
+
+  // Verify that the clone is deep, i.e. parameters of modules are cloned too.
+  torch::NoGradGuard no_grad;
```
Why is this needed?

ejguan replied:
Yes, no need to run autograd for the following function as the test is to verify clone..

zou3519 reviewed (commented):
Not many comments from me but I don't know much about the C++ API

glaringlee commented on torch/csrc/api/include/torch/nn/modules/container/moduledict.h:
```diff
@@ -0,0 +1,185 @@
+  /// Returns the value associated with the given `key`. Throws an exception if
+  /// no such key is stored in the `ModuleDict`. Check contains(key) before
+  /// for a non-throwing way of access.
+  std::shared_ptr<Module>& operator[](const std::string& key) {
```
Why this is a reference instead of a shared_ptr obj?

ejguan replied:
I am trying to keep consistency of our API like `parameterdict` and `orderedDict` for the operator [] and at function to return a value reference.

ejguan replied:
Besides, by using reference, we can assign module to the location. But, it makes me concern about the registration. I think you are right, I need to make it non-reference obj.

glaringlee replied:
@ejguan Yea, there is no much difference here, a reference of shard_ptr is rarely being used (probably can save a counter++ time, I am not sure whether compiler will automatically invoke std::move, if so, then no overhead). Let's do non-reference obj.

glaringlee commented on torch/csrc/api/include/torch/nn/modules/container/moduledict.h:
```diff
@@ -0,0 +1,185 @@
+  /// Returns the value associated with the given `key`. Throws an exception if
+  /// no such key is stored in the `ModuleDict`. Check contains(key) before
+  /// for a non-throwing way of access.
+  const std::shared_ptr<Module>& operator[](const std::string& key) const {
```
Why this is a reference instead of a shared_ptr obj?

ejguan commented on torch/csrc/api/include/torch/nn/modules/container/moduledict.h:
```diff
@@ -0,0 +1,185 @@
+  std::shared_ptr<Module> pop(const std::string& key) {
+    auto module = modules_[key];
+    modules_.erase(key);
+    return module;
```
It's same to `clear()`. I will also make a comment there.

ejguan commented:
And, I am wondering if we need another API of `ptr(const std::string& key)`?  @glaringlee

glaringlee commented:
> And, I am wondering if we need another API of `ptr(const std::string& key)`? @glaringlee

I did not see any usage for ptr at the moment, but you can add it if it is not heavy work.

ejguan committed:
Update on "Implement C++ ModuleDict"


Fixes #45896
- [x] API
* constructor(OrderedDict/vector of pair)
* clear()
* items()
* keys()
* pop(string key)
* update(OrderedDict/vector of pair/ModuleDict)
* values()
* Iterator: begin/end
* size()
* empty()
* contains(string key)
* clone()
* pretty_print(stream)
* operator [string key]
* at(string key)
- [x] Add tests

Differential Revision: [D24872641](https://our.internmc.facebook.com/intern/diff/D24872641)

[ghstack-poisoned]

glaringlee commented on test/cpp/api/moduledict.cpp:
```diff
@@ -0,0 +1,342 @@
+  }
+}
+
+TEST_F(ModuleDictTest, IsCloneable_CUDA) {
```
It seems to me that IsCloneable and IsCloneable_CUDA are pretty much the same except the device.
Can we do something like this so we don't need to maintain 2 pieces of identical code. Sry, I should mention this earlier.
Here is an example:
https://github.com/pytorch/pytorch/blob/e70c50bf59f3ca0d885a716043e2f31aea98be30/test/cpp/api/transformer.cpp#L152

ejguan replied:
Yep, will do. Thanks for the advice!!

glaringlee reviewed (commented):
Just one comment, I should mention earlier, sry.  
Otherwise, all looks good to me.

ejguan committed:
Update on "Implement C++ ModuleDict"


Fixes #45896
- [x] API
* constructor(OrderedDict/vector of pair)
* clear()
* items()
* keys()
* pop(string key)
* update(OrderedDict/vector of pair/ModuleDict)
* values()
* Iterator: begin/end
* size()
* empty()
* contains(string key)
* clone()
* pretty_print(stream)
* operator [string key]
* at(string key)
- [x] Add tests

Differential Revision: [D24872641](https://our.internmc.facebook.com/intern/diff/D24872641)

[ghstack-poisoned]

glaringlee reviewed (approved):
LGTM now. Please sync this latest version on phabricator before landing. thx
